**Author:** Lisa Wallner  
**Description:**  
 

*Hint: If lines are created with support of a Large Language Model or the code is taken from another source, you find following hint at the end of the line: (generated with Microsoft Copilot) or (source: link_to_source)*

In [141]:
import pandas as pd
import json
import matplotlib as plt
import numpy as np

## Dataframe df_llama  
This dataframe gives an overview of the evaluation results from the original and genearted READMES (generated by llama3.1-8b). The scoring was created with three different kind of llms - llama3.1-8b, reka-flash & jamba-1.5-mini.

### Universal Functions

In [142]:
def exclude_repos(df):
    repos_exclude = [('diego-vicente', 'som-tsp'), ('Taniiishk', 'Rock-Paper-Scissors-Game')]
    df = df[~df.set_index(['repo_owner', 'repo_name']).index.isin(repos_exclude)] 

    return df

In [143]:
def remove_cols(df):
    df = df.drop(['readme_g_score', 'readme_g_score_q1_txt', 'readme_g_score_q2_txt', 'readme_g_score_q3_txt', 'readme_g_score_q4_txt', 'readme_g_score_q5_txt', 'readme_o_score', 'readme_o_score_q1_txt', 'readme_o_score_q2_txt', 'readme_o_score_q3_txt', 'readme_o_score_q4_txt', 'readme_o_score_q5_txt'], axis=1)

    return df

In [144]:
def load_data(path):
    with open(path, 'r') as f:
        loaded_data = json.load(f)

    df = pd.DataFrame(data=loaded_data)

    return df

### Model 1

In [145]:
model = 'model1'
path = f'../data/df_score_lama/df_score_{model}_manuel.json'

In [146]:
df_1 = load_data(path)

During the process there are some repositories found which can not be used because of inconsistence in the data. These repository data will be filterd out.  
+ diego-vicente: som-tsp --> GitHub repository has no README file
+ Taniiishk: Rock-Paper-Scissors-Game --> GitHub repository has no license

In [147]:
len(df_1)

201

In [148]:
df_1.head(2)

repo_owner repo_name  \
0              keras-rl  keras-rl   
1  terraform-compliance       cli   

                                      readme_g_score  readme_g_score_q1  \
0  ### "q1": [\n    ##"score": 5##,\n    ##"expla...                  5   
1  ### "q1": [ \n    ##"score": 5##,\n    ##"expl...                  5   

                               readme_g_score_q1_txt  readme_g_score_q2  \
0  The README clearly states that the project is ...                  5   
1  The project's goal is clearly stated in the "D...                  5   

                               readme_g_score_q2_txt  readme_g_score_q3  \
0  The README clearly explains that the project i...                  4   
1  The project is useful because it provides a BD...                  4   

                               readme_g_score_q3_txt  readme_g_score_q4  ...  \
0  The README provides a clear example code to ge...                  2  ...   
1  The README provides a clear step-by-step guide...                  2  ...   

  readme_o_score_q1                              readme_o_score_q1_txt  \
0                 5  The goal of the project is clearly stated in t...   
1                 5  The goal of the project is clearly stated in t...   

  readme_o_score_q2                              readme_o_score_q2_txt  \
0                 5  The project is useful because it provides a se...   
1                 5  The README clearly explains why the project is...   

   readme_o_score_q3                              readme_o_score_q3_txt  \
0                  4  The README provides a clear installation guide...   
1                  4  The README provides a clear installation guide...   

   readme_o_score_q4                              readme_o_score_q4_txt  \
0                  5  The README provides several ways for users to ...   
1                  5  The README provides several ways for users to ...   

   readme_o_score_q5                              readme_o_score_q5_txt  
0                  5  The README clearly states that the project is ...  
1                  5  The README clearly states that the project is ...  

[2 rows x 24 columns]

In [149]:
df_1 = exclude_repos(df_1)

In [150]:
len(df_1)

199

For the final dataframe not all columns are required. Dispensable columns are removed to keep the dataframe clean.

In [151]:
df_1 = remove_cols(df_1)

In [152]:
df_1.describe()

readme_g_score_q1  readme_g_score_q2  readme_g_score_q3  \
count         199.000000         199.000000         199.000000   
mean            4.542714           4.577889           4.316583   
std             0.686769           0.740402           0.581948   
min             1.000000           2.000000           2.000000   
25%             4.000000           4.000000           4.000000   
50%             5.000000           5.000000           4.000000   
75%             5.000000           5.000000           5.000000   
max             5.000000           5.000000           5.000000   

       readme_g_score_q4  readme_g_score_q5  readme_o_score_q1  \
count         199.000000         199.000000         199.000000   
mean            2.135678           2.266332           4.693467   
std             0.701142           0.950503           0.798678   
min             1.000000           1.000000           1.000000   
25%             2.000000           2.000000           5.000000   
50%             2.000000           2.000000           5.000000   
75%             2.000000           2.000000           5.000000   
max             5.000000           5.000000           5.000000   

       readme_o_score_q2  readme_o_score_q3  readme_o_score_q4  \
count         199.000000         199.000000         199.000000   
mean            4.743719           4.296482           4.221106   
std             0.703274           0.743686           1.133225   
min             1.000000           1.000000           1.000000   
25%             5.000000           4.000000           4.000000   
50%             5.000000           4.000000           5.000000   
75%             5.000000           5.000000           5.000000   
max             5.000000           5.000000           5.000000   

       readme_o_score_q5  
count         199.000000  
mean            4.311558  
std             1.264678  
min             1.000000  
25%             4.000000  
50%             5.000000  
75%             5.000000  
max             5.000000

### Model 2

In [153]:
model = 'model2'
path = f'../data/df_score_lama/df_score_{model}_manuel.json'

In [154]:
df_2 = load_data(path)

In [155]:
len(df_2)

201

In [156]:
df_2.head(2)

repo_owner repo_name  \
0              keras-rl  keras-rl   
1  terraform-compliance       cli   

                                      readme_g_score  readme_g_score_q1  \
0   ###\n"q1": [\n    "score": 4,\n    "explanati...                  4   
1   ###\n"q1": [\n    "score": 4,\n    "explanati...                  4   

                               readme_g_score_q1_txt  readme_g_score_q2  \
0  The README clearly states the goal of the proj...                  3   
1  The README clearly states that the project is ...                  3   

                               readme_g_score_q2_txt  readme_g_score_q3  \
0  The usefulness of the project is hinted at by ...                  4   
1  The usefulness of the project is implied by it...                  5   

                               readme_g_score_q3_txt  readme_g_score_q4  ...  \
0  The README offers a clear guide on how to get ...                  2  ...   
1  The README provides clear and detailed instruc...                  2  ...   

  readme_o_score_q1                              readme_o_score_q1_txt  \
0                 5  The README clearly states the goal of the proj...   
1                 5  The README clearly states the goal of the proj...   

  readme_o_score_q2                              readme_o_score_q2_txt  \
0                 4  The project is useful because it provides a st...   
1                 5  The project's usefulness is well-explained thr...   

   readme_o_score_q3                              readme_o_score_q3_txt  \
0                  5  Users can get started with the project very ea...   
1                  4  Users can get started with the project by foll...   

   readme_o_score_q4                              readme_o_score_q4_txt  \
0                  4  Users can get help through several channels: f...   
1                  3  Help resources are mentioned through links to ...   

   readme_o_score_q5                              readme_o_score_q5_txt  
0                  4  The README mentions that the project is mainta...  
1                  3  The README mentions the project maintainer (ee...  

[2 rows x 24 columns]

In [157]:
df_2 = exclude_repos(df_2)

In [158]:
len(df_2)

199

In [159]:
df_2 = remove_cols(df_2)

In [160]:
df_2.describe()

readme_g_score_q1  readme_g_score_q2  readme_g_score_q3  \
count         199.000000         199.000000         199.000000   
mean            4.346734           3.432161           4.663317   
std             0.546224           0.545108           0.514645   
min             2.000000           3.000000           2.000000   
25%             4.000000           3.000000           4.000000   
50%             4.000000           3.000000           5.000000   
75%             5.000000           4.000000           5.000000   
max             5.000000           5.000000           5.000000   

       readme_g_score_q4  readme_g_score_q5  readme_o_score_q1  \
count         199.000000         199.000000         199.000000   
mean            2.331658           3.211055           4.758794   
std             0.513015           0.508194           0.561483   
min             1.000000           2.000000           2.000000   
25%             2.000000           3.000000           5.000000   
50%             2.000000           3.000000           5.000000   
75%             3.000000           3.000000           5.000000   
max             4.000000           5.000000           5.000000   

       readme_o_score_q2  readme_o_score_q3  readme_o_score_q4  \
count         199.000000         199.000000         199.000000   
mean            4.537688           4.693467           4.020101   
std             0.701612           0.620775           0.893070   
min             2.000000           2.000000           1.000000   
25%             4.000000           5.000000           4.000000   
50%             5.000000           5.000000           4.000000   
75%             5.000000           5.000000           5.000000   
max             5.000000           5.000000           5.000000   

       readme_o_score_q5  
count         199.000000  
mean            4.206030  
std             0.786736  
min             2.000000  
25%             4.000000  
50%             4.000000  
75%             5.000000  
max             5.000000

### Model 3

In [161]:
model = 'model3'
path = f'../data/df_score_lama/df_score_{model}_manuel.json'

In [162]:
df_3 = load_data(path)

In [163]:
len(df_3)

201

In [164]:
df_3.head(2)

repo_owner repo_name  \
0              keras-rl  keras-rl   
1  terraform-compliance       cli   

                                      readme_g_score  readme_g_score_q1  \
0   ###\n"q1": [\n    "score": 4,\n    "explanati...                  4   
1   ###\n"q1": [\n    "score": 5,\n    "explanati...                  5   

                               readme_g_score_q1_txt  readme_g_score_q2  \
0  The goal of the project is clearly stated in t...                  5   
1  The goal of the project is clearly stated in t...                  5   

                               readme_g_score_q2_txt  readme_g_score_q3  \
0  The project's usefulness is well-explained in ...                  3   
1  The project's usefulness is evident from its d...                  4   

                               readme_g_score_q3_txt  readme_g_score_q4  ...  \
0  While the installation instructions are clear,...                  3  ...   
1  The installation instructions are clear and co...                  3  ...   

  readme_o_score_q1                              readme_o_score_q1_txt  \
0                 5  The goal of the project is clearly stated in t...   
1                 5  The goal of the project is clearly stated in t...   

  readme_o_score_q2                              readme_o_score_q2_txt  \
0                 5  The project is useful because it provides a li...   
1                 5  The project's usefulness is well explained. It...   

   readme_o_score_q3                              readme_o_score_q3_txt  \
0                  5  The README provides clear instructions on how ...   
1                  5  Getting started is explained in the "Installat...   

   readme_o_score_q4                              readme_o_score_q4_txt  \
0                  5  The README provides information on where users...   
1                  5  Help is readily available through the provided...   

   readme_o_score_q5                              readme_o_score_q5_txt  
0                  5  The README provides information on who maintai...  
1                  5  The project's maintenance and contribution det...  

[2 rows x 24 columns]

In [165]:
df_3 = exclude_repos(df_3)

In [166]:
len(df_3)

199

In [167]:
df_3 = remove_cols(df_3)

In [168]:
df_3.describe()

readme_g_score_q1  readme_g_score_q2  readme_g_score_q3  \
count         199.000000         199.000000         199.000000   
mean            4.793970           4.834171           4.361809   
std             0.474356           0.479518           0.710490   
min             1.000000           1.000000           2.000000   
25%             5.000000           5.000000           4.000000   
50%             5.000000           5.000000           4.000000   
75%             5.000000           5.000000           5.000000   
max             5.000000           5.000000           5.000000   

       readme_g_score_q4  readme_g_score_q5  readme_o_score_q1  \
count         199.000000         199.000000         199.000000   
mean            3.698492           4.241206           4.929648   
std             1.123463           0.949274           0.369378   
min             1.000000           1.000000           1.000000   
25%             3.000000           3.000000           5.000000   
50%             3.000000           5.000000           5.000000   
75%             5.000000           5.000000           5.000000   
max             5.000000           5.000000           5.000000   

       readme_o_score_q2  readme_o_score_q3  readme_o_score_q4  \
count         199.000000         199.000000         199.000000   
mean            4.849246           4.552764           4.281407   
std             0.468597           0.624403           0.969852   
min             1.000000           1.000000           1.000000   
25%             5.000000           4.000000           3.000000   
50%             5.000000           5.000000           5.000000   
75%             5.000000           5.000000           5.000000   
max             5.000000           5.000000           5.000000   

       readme_o_score_q5  
count         199.000000  
mean            4.633166  
std             0.766143  
min             1.000000  
25%             5.000000  
50%             5.000000  
75%             5.000000  
max             5.000000

### Final Dataframe - df_llama

Create Dataframe for the scoring data of 199 GitHub repositories and the three evaluation models lama3.1-8b, reka-flash and jamba-1.5-mini.

In [169]:
cols = ['repo_owner', 'repo_name', 'model_type', 'total_score_g', 'score_q1_g', 'score_q2_g', 'score_q3_g', 'score_q4_g', 'score_q5_g', 'total_score_o', 'score_q1_o', 'score_q2_o', 'score_q3_o', 'score_q4_o', 'score_q5_o']
df_llama = pd.DataFrame(columns=cols)

In [170]:
dfs = [df_1, df_2, df_3]
model_types = ['llama3.1-8b', 'reka-flash', 'jamba-1.5-mini']

In [173]:
for idx, df in enumerate(dfs):
    model_type = model_types[idx]
    
    for _,row in df.iterrows():
        new_row = {
            'repo_owner': row['repo_owner'], 
            'repo_name': row['repo_name'], 
            'model_type': model_type, 
            'total_score_g': np.nan, 
            'score_q1_g': row['readme_g_score_q1'], 
            'score_q2_g': row['readme_g_score_q2'], 
            'score_q3_g': row['readme_g_score_q3'], 
            'score_q4_g': row['readme_g_score_q4'], 
            'score_q5_g': row['readme_g_score_q5'], 
            'total_score_o': np.nan, 
            'score_q1_o': row['readme_o_score_q1'], 
            'score_q2_o': row['readme_o_score_q2'], 
            'score_q3_o': row['readme_o_score_q3'], 
            'score_q4_o': row['readme_o_score_q4'], 
            'score_q5_o': row['readme_o_score_q5']
    }
        df_llama = pd.concat([df_llama, pd.DataFrame([new_row])], ignore_index=True)

Calculate the total score for the genearted and orginal README's.

In [185]:
df_llama['total_score_g'] = df_llama[['score_q1_g', 'score_q2_g', 'score_q3_g', 'score_q4_g', 'score_q5_g']].sum(axis=1)
df_llama['total_score_o'] = df_llama[['score_q1_o', 'score_q2_o', 'score_q3_o', 'score_q4_o', 'score_q5_o']].sum(axis=1)

In [ ]:
# df_llama['total_score_o'][df_llama['model_type'] == 'llama3.1-8b'].value_counts()

total_score_o
24    104
23     21
25     17
21     11
20      9
19      8
18      7
22      5
13      5
16      3
17      2
14      2
9       1
12      1
11      1
6       1
15      1
Name: count, dtype: int64